src/components/header/EmptyThreadIndicator.js

```diff
@@ -126,4 +75,72 @@ class EmptyThreadIndicator extends PureComponent<Props> {
   }
 }
 
+/**
+ * Define this outside of the class so that it's easily testable. The internals
+ * are a little complicated with the math, but the test file should have some
+ * pretty clear explanations of the requirements:
+ * src/test/components/EmptyThreadIndicator.test.js
+ */
+export function getIndicatorPositions(
+  props: Props
+): {|
+  startup: SyntheticCssDeclarations | null,
+  shutdown: SyntheticCssDeclarations | null,
+  emptyBufferStart: SyntheticCssDeclarations | null,
+|} {
+  const {
+    rangeStart,
+    rangeEnd,
+    width,
+    unfilteredSamplesRange,
+    interval,
+    thread: { processShutdownTime, registerTime, unregisterTime },
+  } = props;
+  const rangeLength = rangeEnd - rangeStart;
+  const xPixelsPerMs = width / rangeLength;
+  const threadEndTime =
+    unregisterTime === null ? processShutdownTime : unregisterTime;
+
+  // Did this thread startup in this time range?
+  let startup = null;
+  if (registerTime > rangeStart) {
+    startup = {
+      left: 0,
+      width: Math.min(width, (registerTime - rangeStart) * xPixelsPerMs),
+    };
+  }
+
+  // Did this thread shut down in this time range?
+  let shutdown = null;
+  if (threadEndTime !== null && threadEndTime < rangeEnd) {
+    shutdown = {
+      right: 0,
+      width: Math.min(width, (rangeEnd - threadEndTime) * xPixelsPerMs),
+    };
+  }
+
+  let emptyBufferStart = null;
+  if (
+    // Threads could have no samples, and therefore no range.
+    unfilteredSamplesRange !== null &&
```

Comment:
What do we do if `unfilteredSamplesRange` is null ? Do we assume this never happens ?

Reply:
Hmm... that would mean we literally have no samples for a thread. I'm not sure what this means as far as visualizing it, because we're telling people that the *start* of their buffer is empty and listing the reasons. This would be a different set of reasons. I would be fine just leaving it blank with no annotations.